libs/pika/async_cuda/include/pika/async_cuda/then_with_stream.hpp

```diff
@@ -0,0 +1,627 @@
+    /// argument as the first argument. This can only be called on a sender with
+    /// a completion scheduler that is cuda_scheduler. The handle is
+    /// thread-local and f may not yield a pika thread until after the handle
+    /// has been used the last time by f.
+    inline constexpr struct then_with_cublas_t final
+    {
+        template <typename S, typename F>
+        constexpr PIKA_FORCEINLINE auto operator()(
+            S&& s, F&& f, cublasPointerMode_t pointer_mode) const
+            -> decltype(detail::then_with_cuda_pool(PIKA_FORWARD(S, s),
+                detail::cublas_handle_callable<F>{
+                    PIKA_FORWARD(F, f), pointer_mode}))
+        {
+            return detail::then_with_cuda_pool(PIKA_FORWARD(S, s),
+                detail::cublas_handle_callable<F>{
+                    PIKA_FORWARD(F, f), pointer_mode});
+        }
+
+        template <typename F>
+        constexpr PIKA_FORCEINLINE auto operator()(
+            F&& f, cublasPointerMode_t pointer_mode) const
+            -> decltype(
+                pika::execution::experimental::detail::partial_algorithm<
+                    then_with_cublas_t, F, cublasPointerMode_t>{
+                    PIKA_FORWARD(F, f), pointer_mode})
+        {
+            return pika::execution::experimental::detail::partial_algorithm<
+                then_with_cublas_t, F, cublasPointerMode_t>{
+                PIKA_FORWARD(F, f), pointer_mode};
+        }
+    } then_with_cublas{};
+
+#if defined(PIKA_HAVE_CUDA)
+    /// Attach a continuation to run f with an additional cuSOLVER handle.
+    ///
+    /// Attaches a continuation to the given sender which will call f with the
+    /// arguments sent by the given sender with an additional cusolverDnHandle_t
+    /// argument as the first argument. This can only be called on a sender with
+    /// a completion scheduler that is cuda_scheduler. The handle is
+    /// thread-local and f may not yield a pika thread until after the handle
+    /// has been used the last time by f.
+    inline constexpr struct then_with_cusolver_t final
+    {
+        template <typename S, typename F>
+        constexpr PIKA_FORCEINLINE auto operator()(S&& s, F&& f) const
+            -> decltype(detail::then_with_cuda_pool(PIKA_FORWARD(S, s),
+                detail::cusolver_handle_callable<F>{PIKA_FORWARD(F, f)}))
+        {
+            return detail::then_with_cuda_pool(PIKA_FORWARD(S, s),
+                detail::cusolver_handle_callable<F>{PIKA_FORWARD(F, f)});
+        }
+
+        template <typename F>
+        constexpr PIKA_FORCEINLINE auto operator()(F&& f) const -> decltype(
+            pika::execution::experimental::detail::partial_algorithm<
+                then_with_cusolver_t, F>{PIKA_FORWARD(F, f)})
+        {
+            return pika::execution::experimental::detail::partial_algorithm<
+                then_with_cusolver_t, F>{PIKA_FORWARD(F, f)};
+        }
+    } then_with_cusolver{};
+#endif
```

Comment:
Note to self: Move the generic `then_with_anything_cuda_thing` from DLA-Future into here as well.